channels/jaicp/src/main/kotlin/com/justai/jaicf/channel/jaicp/dto/JaicpDialerAPI.kt

```diff
@@ -0,0 +1,113 @@
+package com.justai.jaicf.channel.jaicp.dto
+
+import com.justai.jaicf.channel.jaicp.JSON
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.json.JsonElement
+
+@Serializable
+class JaicpDialerAPI {
+
+    private var callResult: String? = null
+    private var callResultPayload: String? = null
+    private var reportData: MutableMap<String, CallReportData> = mutableMapOf()
+    private var redial: RedialData? = null
+
+    /**
+     * Reports data to be stored in .xsls report.
+     *
+     * @param value - value in cell in .xsls report
+     * @param order - optional order for column
+     * */
+    @Serializable
+    internal class CallReportData(
+        private val value: String?,
+        private val order: Int?
+    )
+
+    /**
+     * An object sent to schedule a redial.
+     *
+     * @property startDateTime unix timestamp (UTC-0 epoch milliseconds) to start attempting to redial a client
+     * @property finishDateTime unix timestamp (UTC-0 epoch milliseconds) to end attempting to redial a client
+     * @property allowedDays list of [DayOfWeek] allowed days
+     * @property localTimeFrom local time interval start attempting to redial. E.g. 16:20
+     * @property localTimeTo local time interval end attempting to redial. E.g. 23:59
+     * @property retryIntervalInMinutes interval between redial attempts
+     * @property maxAttempts max number of attempts to call client
+     * */
+    @Serializable
+    data class RedialData(
+        val startDateTime: Long? = null,
+        val finishDateTime: Long? = null,
+        val allowedDays: List<DayOfWeek> = emptyList(),
+        val localTimeFrom: String? = null,
+        val localTimeTo: String? = null,
+        val retryIntervalInMinutes: Int? = null,
+        val maxAttempts: Int? = null
+    )
+
+    internal fun redial(data: RedialData) {
+        checkStartFinishTime(data)
+        checkLocalTime(data)
+        checkRetryAndInterval(data)
+        redial = data
+    }
+
+    internal fun report(header: String, data: CallReportData) {
+        reportData[header] = data
+    }
+
+    internal fun result(result: String?, resultPayload: String?) {
+        callResult = result
+        callResultPayload = resultPayload
+    }
+
+    internal fun getApiResponse(): JsonElement {
+        return JSON.toJson(serializer(), this)
+    }
+}
+
+enum class DayOfWeek(val value: String) {
+    MON("mon"),
+    TUE("tue"),
+    WED("wed"),
+    THU("thu"),
+    FRI("fri"),
+    SAT("sat"),
+    SUN("sun")
+}
+
+private fun checkStartFinishTime(data: JaicpDialerAPI.RedialData) {
+    val st = data.startDateTime
+    val fin = data.finishDateTime
+    if (st != null && fin != null) {
```

Comment:
I don't think we should check for this kind of errors. We should check only most primitive and obvious.

Generally, If `fin < System.currentTimeMillis() (Plus some time for network latency and dialer schedule)`, redial won't be scheduled because redial finish time had passed. And I don't see a proper solution for this.

For example, your fin value can be correct by framework (current time + 1 min), but dialer has internal call pool and there were no place in pool for minute, therefore scheduled call won't redial.

Reply:
That's true about really complex checks, but we can at least check that `st` and `fin` are greater than 0

Reply:
It makes sense, but throwing this error may mislead user about logic inside dialer.

`st` and `fin` being positive numbers is not a requirement of dialerAPI. The requirement is that st/fin must be a **reasonable** timestamp, and we cannot check it, and I guess we should not throw misleading errors.

Reply:
Well, we can say the same about `st <= fin` check, can't we?
I don't really figure out, why we put some little checks on our own and discard others?

Maybe I misunderstand you, what do you mean by saying that:
> st and fin being positive numbers is not a requirement of dialerAPI

Is it possible to pass a negative timestamp to the dialerAPI and don't get an error?
If not, then it is a requirement of dialerAPI

Reply:
> Well, we can say the same about st <= fin check, can't we?

Well, yes. But in this case we can provide accurate error message which will tell where user is wrong exaclty.

The only requirement of dialerAPI is that start/fin should make sense. **We only check things which are possible to check on our own**, for example, HHmm formatting, maxAttempts or retryInterval.

**What do we want to check**: that with given start/finish time we will redial in provided interval.
**How we're gonna check this:**  no idea, this particular check is impossible for us. 

So we decide to check some other thing, which does not solve our problem and does mislead user about internal logic. Given an error like `start and fin should be greater then zero` user will still have no idea why his params `start=10, fin=20` do not work.

Reply:
Well, again, there is no difference with `st <= fin`.

> Given an error like "start and fin should be greater then zero" user will still have no idea why his params start=10, fin=20 do not work.

It will say to the user, that params `start = -5` is not correct.

The same for `st <= fin`: user will still have no idea why params `start = 10, fin = 20` do not work, but he definitely will know that `start = 20, fin = 10` do not work.

Reply:
But difference is still present: error message for `st <= fin` is accurate and explains why user is wrong in this particular piece of code.

Reply:
What is the problem with `Time cannot be negative`?

Reply:
It sounds like "redial me please yesterday". It's an idiot-check, but not a check which should be in source code.

You can just mess up parameters order when invoking a reaction, and you have st < fin check for it. But do we need to check  user code like `Instant.ofEpochMillis(-5)`?

Reply:
What about `ArrayIndexOutOfBoundsException`? I'd not say that it is an idiot check.

>  But do we need to check user code like Instant.ofEpochMillis(-5)?

Yes we do, as we don't check Instant, we check Long, and it's pretty easy to forget to add such a check if we change the constructor

Reply:
`ArrayIndexOutOfBoundsException` is quite different thing. Your array is dynamic in some way, but not time in this case.

I refuse to add this check because it conflicts with my understanding of programming and writing of a framework. If we do this, we should error-check a hundred things in our framework, which just MAY occur if user don't understand what he/she's doing. 

You can add this commit if you want, I just don't want to be an author of this check.

Reply:
>  Your array is dynamic in some way

In any way `array[-1]` is as much "idiotic" as `startTime = -1`

>  If we do this, we should error-check a hundred things in our framework, which just MAY occur if user don't understand what he/she's doing.

And you are absolutely right. That's how it should be done.

Reply:
If you write in Kotlin and not Python, then yes, array[-1] is idiotic. 
If you want to schedule a redial (an operation in future) use negative time (past) is idiotic.

 > And you are absolutely right. That's how it should be done.

Do we assume that our users are cats who have no conception of `redial`?

Reply:
Do we assume, that our users live in another universe, where start time could be bigger than finish time?

Don't we assume, that our users can use computable values in parameters, that can become negative because of overflow, for example?

Reply:
What crazy mess should a man write before redial to suffer from long overflow? 
If a man writes that, no checks will help him.